_posts/2017-10-06-rust-blog-series-2.md

```diff
@@ -0,0 +1,1144 @@
+Most ORMs use a [*prepared statement cache*] which stores our queries to avoid re-parsing and processing.
+Our database backend might be processing statements from multiple clients concurrently
+and can be slowed down if it unnecessarily repeats operations.
+In fact, database connections aren't the only type of connections that benefit from being reused.
+Most browsers reuse connections for HTTP requests because even the cost of a TCP handshake can impact performance.
+
+For these reasons,
+we don't really want every single visitor to our blog opening up a new database connection everytime they hit a route.
+What we need is a [Connection Pool].
+Connection pools are a *cache* / store of connections that stick around
+and can be reused to help enhance the performance of executing commands on our database.
+When a connection is needed, it is plucked from the pool, used,
+and then returned back to be available for another connection.
+
+[*prepared statement cache*]: http://www.theserverside.com/news/1365244/Why-Prepared-Statements-are-important-and-how-to-use-them-properly
+[remember reading somewhere]: https://stackoverflow.com/questions/34303678/database-connection-expensive-to-create
+[Connection Pool]: https://en.wikipedia.org/wiki/Connection_pool
+
+[r2d2] is a generic connection pool library for Rust and it happens to have an adapter for Diesel, [r2d2-diesel]!
+We'll be using the both libraries to manage the connection pool to our Postgres database.
+
+[r2d2]: https://github.com/sfackler/r2d2
+[r2d2-diesel]: https://github.com/diesel-rs/r2d2-diesel
+
+Our immediate goal:
+- set up a database connection pool so Rocket may pass around connections to Postgres
+
+Be forewarned, I'm going to make an arbitrary design decision here and put all our connection logic into `src/lib.rs`.
+This will leave `src/bin/main.rs` to hold all the Rocket initialization / setup.
+Consequently, all of our app business logic will be pulled together in `lib.rs` from other modules,
+leaving most of our files focused and arranged by features.
+
+Remember `src/schema.rs` and `src/models.rs`?
+They will both be re-exported by our `lib` so various parts of our application may make use of them.
+
+I should cover [Rocket Request Guards] and [Managed State] before we get started,
+because they're fundamental to any Rocket application - even at this stage.
+Request guards are *similar* to middleware, in that they intercept requests.
+A request guard is mainly used for request validation and can force redirection based on their outcome.
+The validation policy can be built in, like we see from standard rocket request objects,
+or we can implement our own custom validation policies like we will for passing the connection pool.
+
+Any type that implements the `FromRequest` trait is a request guard.
+It is important to note that we won't be implementing this trait directly on our database connection.
+Our database connection is *state* that will passed around,
```

Comment:
I think this sentence is missing some words: `Our database connection is a *state* that will be passed around,`